Rpak Pilot Model Importing (#159)
R2Northstar/ModdingDocs: docs/source/guides/tools/rpakpilotmodding.rst

.. _importpilotmodel:

Rpak Pilot Model Importing
=====

Tools:
^^^^^

Download each tool. 

.. _Blender: https://www.blender.org/

.. _Blender Source Tools: http://steamreview.org/BlenderSourceTools/

.. _Titanfall VPK Tools: https://retryy.gitbook.io/tf2/Wiki/Tools/vpk-tools

.. _Crowbar: https://steamcommunity.com/groups/CrowbarTool

.. _Legion+: https://github.com/r-ex/LegionPlus

.. _RePak: https://github.com/r-ex/RePak

.. _mdlshit: https://github.com/headassbtw/mdlshit-binaries

.. _Text Editors: https://retryy.gitbook.io/tf2/Wiki/Tools/general-pc-tools#text-editing

- `Text Editors`_ - Just a text editor

- `Titanfall VPK Tools`_ - VPK viewer, extractor, and repacker.
            "Titanfall VPK Tool" is the vpk tool used. Made by cra0.
  

> - `Crowbar`_ - Decompile, recompile, and a ton of other model and source asset stuff

Reviewer: I would prefer if a different vpk tool was recommended here, as they are generally considered better these days (Crowbar can do vpk extraction for titanfall 2 properly, so you could just recommend Crowbar for it to avoid having too many required tools)

This change would however require a rewriting/rewording of the section dedicated to extracting the model from the vpk

Author: I never recommended a specific vpk tool.

I will take another look in VPK tools to see what is the overall best, but I just continued with cra0's VPKTool because it's popular use in tutorials, it's ability to unpack AND repack. 

If I choose Crowbar, I'll rewrite the necessary content.

  
- `Legion+`_  - Texture, audio, shaderset, and material exporter
               "Normal" Legion doesn't work
  
- `RePak`_ - Create Apex and Titanfall 2 .rpak for textures
  
- `mdlshit`_ - Magically convert .mdl to "titanfall 2" .mdl 

Exporting the pilot model from the vpk
^^^^^

Locate your Titanfall 2 game folder
-----

For Origin you should find it in this directory ``Origin Games\Titanfall2``.

For Steam you should find it in this directory ``Steam\steamapps\common\Titanfall2``.

Open Titanfall VPK Tool
-----

At the top left press the ``Open`` button.
Open the vpk folder then open ``englishclient_mp_common.bsp.pak000_dir.vpk``,
Expand the ``models`` folder, then the ``humans`` folder, then click on the ``pilots`` folder.

The files have different names that correspond to their pilot classes. Below is a clarifying table to help.

====================    ===== 
VPK .mdl name           Name
====================    =====
pilot_light_jester      Stim
pilot_light_ged         Phase Shift
pilot_medium_stalker    Holo Pilot
pilot_medium_reaper     Pulse Blade
pilot_medium_geist      Grapple
pilot_heavy_roog        A-Wall
pilot_heavy_drex        Cloak

> ====================    =====

Reviewer: This should probably also specify that _f models are the female ones and _m models are the male ones, for completeness and to avoid any potential confusion

Author: Will add this!

.. image:: /_static/importpilotmodel/ss2-pilotBlenderMatl.png
   :align: center
   :class: screenshot

Once you have made all the changes you wanted to, select ``Scene Properties``, ``Source Engine Export``, then set the export format to ``SMD`` and set your export path.

.. image:: /_static/importpilotmodel/ss3-BlenderQcModifier.png
   :align: center
   :class: screenshot

Copy the qc file to the same location where you exported the model.

Modifying the POV model
^^^^^

- Select ``File`` in Blender, then select ``Save Copy``
- Choose any name, save the copy, then open the copy
- Delete everything except for the body, the body collection, and the skeleton
- Add ``pov_`` to the beginning of your body model and armature
- Repeat the process of exporting the model with ``Source Engine Export``

Using Legion+ to view paths
^^^^^

Open `Legion+`_
-----

Select ``Load File``
Follow the directory that matches your launcher.

1. Steam: ``Steam\steamapps\common\Titanfall2\r2\paks\Win64\common.rpak`` or
2. Origin: ``Origin Games\Titanfall2\r2\paks\Win64\common.rpak``

Type the file name of the pilot into the search bar to view **texture and material paths**.

> View file paths to create paths in RePak

Reviewer: This should advise to turn on "Show full asset paths" in Legion+ settings, which I believe is off by default

Author: I noticed that but this was just an .rst translation so I kept it "as is".  Considering the amount of adjustments and rewriting, I will now add your suggestion.

Recompiling the model
^^^^^

Install compatible Source game
----

You need a compatible Source game installed when using Crowbar.
Compatible Source games in order of preference:

Compatible Source games
#####

1. Portal 2
2. Alien Swarm
3. Source FilmMaker

Open `Crowbar`_
-----

Select ``Compile``
Click on the Browse button next to ``QC input`` then locate and select your modified model.
Next, Click on the Browse button next to ``Output to`` then choose a location to output the model to.

> Making a Pak file
> ^^^^^
> 
> Follow this `tutorial <https://r2northstar.readthedocs.io/en/latest/guides/rpakmodding.html>`_ to create a pak file

Reviewer: ```suggestion
Making an RPak file
^^^^^

Follow this `tutorial <https://r2northstar.readthedocs.io/en/latest/guides/rpakmodding.html>`_ to create a RPak file
```
Just for consistency, RPak shouldn't be shortened to just Pak (some people might get confused)

Author: Didn't notice. Will use this commit.

Making changes to the qc file (Optional)
^^^^^

Open any text editor 
If you want to prevent camos from being used on your model, delete any textures that have skn31 in the texture group

.. figure:: /_static/importpilotmodel/ss4-qcFileChanges.png
   :align: center

.. figure:: /_static/importpilotmodel/ss5-qcFileChanges2.png
    :align: center

Creating a Northstar mod
^^^^^

Follow the "Getting Started" guide to create a Northstar mod
Create a folder that matches the below file structure

.. note:: Only create the ``weapons`` folder with everything else inside it if you made a pov model

.. Directory Structure for Northstar Mod
:: 

    AuthorName.Mod
    ├──mod
    |   └──models
    |       └──humans
    |           └──pilots

>     |               └──customModel

Reviewer: ```suggestion
    |               └──customModel.mdl
```
Other files have their file extension, so should the model

Author: 100%. Will be using your commit.

    ├──paks
    |   ├──rpak.json
    |   ├──example.rpak

>     |   └──preloadexample.rpak

Reviewer: ```suggestion
```
The tutorial never mentioned making a preload rpak (and they shouldn't be needed anymore either afaik?)

Author: Got it. Using your commit.

    └──mod.json

Credits: 
^^^^^

Base guide from Harb. 

> Edited by Northstar.

Reviewer: Any chance the base guide could be linked here?

Author: Sure.

I'll add the original google link, and experiment with the file sharing on the docs.
